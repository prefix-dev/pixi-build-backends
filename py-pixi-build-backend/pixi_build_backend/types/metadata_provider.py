"""
Python metadata provider types and protocols.
"""

from typing import Optional, Protocol


class MetadataProvider(Protocol):
    """
    Protocol for providing package metadata.
    
    This protocol defines the interface that metadata providers must implement
    to supply package metadata like name, version, description, etc.
    """

    def name(self) -> Optional[str]:
        """Return the package name or None if not available."""
        return None

    def version(self) -> Optional[str]:
        """Return the package version or None if not available."""
        return None

    def homepage(self) -> Optional[str]:
        """Return the package homepage URL or None if not available."""
        return None

    def license(self) -> Optional[str]:
        """Return the package license or None if not available."""
        return None

    def license_file(self) -> Optional[str]:
        """Return the path to the license file or None if not available."""
        return None

    def summary(self) -> Optional[str]:
        """Return a short package summary or None if not available."""
        return None

    def description(self) -> Optional[str]:
        """Return the package description or None if not available."""
        return None

    def documentation(self) -> Optional[str]:
        """Return the documentation URL or None if not available."""
        return None

    def repository(self) -> Optional[str]:
        """Return the repository URL or None if not available."""
        return None