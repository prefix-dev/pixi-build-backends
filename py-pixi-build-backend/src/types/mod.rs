mod config;
mod generated_recipe;
mod platform;
mod project_model;
mod python_params;

pub use generated_recipe::{PyGenerateRecipe, PyGeneratedRecipe, PyVecString};
pub use platform::PyPlatform;
pub use project_model::PyProjectModelV1;

pub use config::PyBackendConfig;
pub use python_params::PyPythonParams;
