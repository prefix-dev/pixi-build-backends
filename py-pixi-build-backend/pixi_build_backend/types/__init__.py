# Re-export all the types
from .metadata_provider import MetadataProvider as MetadataProvider