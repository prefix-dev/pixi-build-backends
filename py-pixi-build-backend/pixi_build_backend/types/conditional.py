from typing import Optional
from typing_extensions import TypeAlias
from pixi_build_backend.pixi_build_backend import (
    PyConditionalString,
    PyListOrItemString,
    PyListOrItemPackageDependency,
    PyConditionalPackageDependency,
)


# Type aliases for FFI types
ListOrItemString: TypeAlias = PyListOrItemString
ListOrItemPackageDependency: TypeAlias = PyListOrItemPackageDependency


class ConditionalString:
    _inner: PyConditionalString

    def __init__(self, condition: str, then: ListOrItemString, else_: Optional[ListOrItemString]) -> None:
        else_ = else_ if else_ is not None else ListOrItemString([])
        self._inner = PyConditionalString(condition, then, else_)

    @property
    def condition(self) -> str:
        """Get the condition string."""
        return self._inner.condition

    @property
    def then_value(self) -> "ListOrItemString":
        """Get the then value."""
        return self._inner.then_value
        # return result

    @property
    def else_value(self) -> ListOrItemString:
        """Get the else value."""
        return self._inner.else_value

    def __str__(self) -> str:
        return str(self._inner)

    def __eq__(self, other: object) -> bool:
        """Check equality."""
        if not isinstance(other, ConditionalString):
            return False
        return self._inner == other._inner


class ConditionalPackageDependency:
    _inner: PyConditionalPackageDependency

    def __init__(self, condition: str, then: ListOrItemPackageDependency, else_: ListOrItemPackageDependency) -> None:
        self._inner = PyConditionalPackageDependency(condition, then, else_)

    @property
    def condition(self) -> str:
        """Get the condition string."""
        return self._inner.condition

    @condition.setter
    def condition(self, value: str) -> None:
        """Set the condition string."""
        self._inner.condition = value

    @property
    def then_value(self) -> ListOrItemPackageDependency:
        """Get the then value."""
        return self._inner.then_value

    @property
    def else_value(self) -> ListOrItemPackageDependency:
        """Get the else value."""
        return self._inner.else_value
