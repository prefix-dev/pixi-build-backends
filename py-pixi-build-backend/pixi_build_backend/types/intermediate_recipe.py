from typing import Optional, List, Dict
from pathlib import Path
from pixi_build_backend.pixi_build_backend import (
    PyIntermediateRecipe,
    PyPackage,
    PyBuild,
    PyConditionalRequirements,
    PyAbout,
    PyExtra,
    PyScript,
    PyPython,
    PyNoArchKind,
    PyValueString,
    PyValueU64,
    PySource,
    PyUrlSource,
    PyPathSource,
    PyPackageSpecDependencies,
    PyItemPackageDependency,
    PyItemString,
)
from pixi_build_backend.types.platform import Platform
from pixi_build_backend.types.requirements import PackageDependency


ConditionalListPackageDependency = List["ItemPackageDependency"]
ConditionalListString = List["ItemString"]


class IntermediateRecipe:
    """An intermediate recipe wrapper."""

    _inner: PyIntermediateRecipe

    def __init__(self) -> None:
        self._inner = PyIntermediateRecipe()

    @property
    def package(self) -> "Package":
        """Get the package information."""
        return Package._from_inner(self._inner.package)
    
    @package.setter
    def package(self, value: "Package") -> None:
        """Set the package information."""
        self._inner.package = value._inner


    @property
    def build(self) -> "Build":
        """Get the build configuration."""
        return Build._from_inner(self._inner.build)
    
    @build.setter
    def build(self, value: "Build") -> None:
        """Set the build configuration."""
        self._inner.build = value._inner

    @property
    def requirements(self) -> "ConditionalRequirements":
        """Get the requirements configuration."""
        return ConditionalRequirements._from_inner(self._inner.requirements)
    
    @requirements.setter
    def requirements(self, value: "ConditionalRequirements") -> None:
        """Set the requirements configuration."""
        self._inner.requirements = value._inner

    @property
    def about(self) -> Optional["About"]:
        """Get the about information."""
        inner_about = self._inner.about
        return About._from_inner(inner_about) if inner_about else None

    @property
    def extra(self) -> Optional["Extra"]:
        """Get the extra information."""
        inner_extra = self._inner.extra
        return Extra._from_inner(inner_extra) if inner_extra else None

    def __repr__(self) -> str:
        return self._inner.__repr__()

    @classmethod
    def _from_inner(cls, inner: PyIntermediateRecipe) -> "IntermediateRecipe":
        """Create an IntermediateRecipe from a PyIntermediateRecipe."""
        instance = cls()
        instance._inner = inner
        return instance

    @staticmethod
    def from_yaml(yaml: str) -> "IntermediateRecipe":
        """
        Create an IntermediateRecipe from a YAML string.

        Parameters
        ----------
        yaml : str
            The YAML string representing the recipe.

        Returns
        -------
        IntermediateRecipe
            The constructed IntermediateRecipe object.

        Examples
        --------
        ```python
        >>> yaml_str = "package:\\n  name: test\\n  version: 1.0.0"
        >>> recipe = IntermediateRecipe.from_yaml(yaml_str)
        >>> recipe.package.name.get_concrete()
        'test'
        >>>
        ```
        """
        return IntermediateRecipe._from_inner(PyIntermediateRecipe.from_yaml(yaml))

    def to_yaml(self) -> str:
        """
        Convert the IntermediateRecipe to a YAML string.

        Returns
        -------
        str
            The YAML representation of the IntermediateRecipe.

        Examples
        --------
        ```python
        >>> recipe = IntermediateRecipe()
        >>> yaml_output = recipe.to_yaml()
        >>> isinstance(yaml_output, str)
        True
        >>>
        ```
        """
        return self._inner.to_yaml()
    
    def __str__(self) -> str:
        """
        Get the string representation of the IntermediateRecipe.

        Returns
        -------
        str
            The YAML representation of the IntermediateRecipe.

        """
        return str(self._inner)


class Package:
    """A package wrapper."""

    _inner: PyPackage

    def __init__(self, name: str, version: str):
        self._inner = PyPackage(ValueString.concrete(name)._inner, ValueString.concrete(version)._inner)

    @property
    def name(self) -> "ValueString":
        """Get the package name."""
        return ValueString._from_inner(self._inner.name)
    
    @name.setter
    def name(self, value: str) -> None:
        """Set the package name."""
        self._inner.name = ValueString(value)._inner

    @property
    def version(self) -> "ValueString":
        """Get the package version."""
        return ValueString._from_inner(self._inner.version)

    @version.setter
    def version(self, value: "ValueString") -> None:
        """Set the package version."""
        self._inner.version = value._inner

    @classmethod
    def _from_inner(cls, inner: PyPackage) -> "Package":
        """Create a Package from a PyPackage."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance


class Build:
    """A build configuration wrapper."""

    _inner: PyBuild

    def __init__(self) -> None:
        self._inner = PyBuild()

    @property
    def number(self) -> Optional["ValueU64"]:
        """Get the build number."""
        inner_number = self._inner.number
        return ValueU64._from_inner(inner_number) if inner_number else None

    @number.setter
    def number(self, value: Optional["ValueU64"]) -> None:
        """Set the build number."""
        self._inner.number = value._inner if value else None

    @property
    def script(self) -> "Script":
        """Get the build script."""
        return Script._from_inner(self._inner.script)

    @script.setter
    def script(self, value: "Script") -> None:
        """Set the build script."""
        self._inner.script = value._inner

    @property
    def noarch(self) -> Optional["NoArchKind"]:
        """Get the noarch kind."""
        inner_noarch = self._inner.noarch
        return NoArchKind._from_inner(inner_noarch) if inner_noarch else None

    @noarch.setter
    def noarch(self, value: Optional["NoArchKind"]) -> None:
        """Set the noarch kind."""
        self._inner.noarch = value._inner if value else None

    @property
    def python(self) -> "Python":
        """Get the Python configuration."""
        return Python._from_inner(self._inner.python)

    @python.setter
    def python(self, value: "Python") -> None:
        """Set the Python configuration."""
        self._inner.python = value._inner

    @classmethod
    def _from_inner(cls, inner: PyBuild) -> "Build":
        """Create a Build from a PyBuild."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance


class Script:
    """A script wrapper."""

    _inner: PyScript

    def __init__(self, content: List[str], env: Optional[Dict[str, str]] = None):
        self._inner = PyScript(content, env)

    @property
    def content(self) -> List[str]:
        """Get the script content."""
        return self._inner.content

    @content.setter
    def content(self, value: List[str]) -> None:
        """Set the script content."""
        self._inner.content = value

    @property
    def env(self) -> Dict[str, str]:
        """Get the environment variables."""
        return self._inner.env

    @env.setter
    def env(self, value: Dict[str, str]) -> None:
        """Set the environment variables."""
        self._inner.env = value

    @property
    def secrets(self) -> List[str]:
        """Get the secrets."""
        return self._inner.secrets

    @secrets.setter
    def secrets(self, value: List[str]) -> None:
        """Set the secrets."""
        self._inner.secrets = value

    @classmethod
    def _from_inner(cls, inner: PyScript) -> "Script":
        """Create a Script from a PyScript."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance


class Python:
    """A Python configuration wrapper."""

    _inner: PyPython

    def __init__(self, entry_points: List[str]):
        self._inner = PyPython(entry_points)

    @property
    def entry_points(self) -> List[str]:
        """Get the entry points."""
        return self._inner.entry_points

    @entry_points.setter
    def entry_points(self, value: List[str]) -> None:
        """Set the entry points."""
        self._inner.set_entry_points(value)

    @classmethod
    def _from_inner(cls, inner: PyPython) -> "Python":
        """Create a Python from a PyPython."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance
    
    def __str__(self) -> str:
        """
        Get the string representation of the Python configuration.

        Returns
        -------
        str
            The string representation of the entry points.
        """
        return str(self._inner)


class NoArchKind:
    """A NoArch kind wrapper."""

    _inner: PyNoArchKind

    @classmethod
    def python(cls) -> "NoArchKind":
        """
        Create a Python NoArch kind.

        Examples
        --------
        ```python
        >>> kind = NoArchKind.python()
        >>> kind.is_python()
        True
        >>> kind.is_generic()
        False
        >>>
        ```
        """
        instance = cls.__new__(cls)
        instance._inner = PyNoArchKind.python()
        return instance

    @classmethod
    def generic(cls) -> "NoArchKind":
        """
        Create a Generic NoArch kind.

        Examples
        --------
        ```python
        >>> kind = NoArchKind.generic()
        >>> kind.is_generic()
        True
        >>> kind.is_python()
        False
        >>>
        ```
        """
        instance = cls.__new__(cls)
        instance._inner = PyNoArchKind.generic()
        return instance

    def is_python(self) -> bool:
        """
        Check if this is a Python NoArch kind.

        Examples
        --------
        ```python
        >>> NoArchKind.python().is_python()
        True
        >>> NoArchKind.generic().is_python()
        False
        >>>
        ```
        """
        return self._inner.is_python()

    def is_generic(self) -> bool:
        """
        Check if this is a Generic NoArch kind.

        Examples
        --------
        ```python
        >>> NoArchKind.generic().is_generic()
        True
        >>> NoArchKind.python().is_generic()
        False
        >>>
        ```
        """
        return self._inner.is_generic()

    @classmethod
    def _from_inner(cls, inner: PyNoArchKind) -> "NoArchKind":
        """Create a NoArchKind from a PyNoArchKind."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance


class ValueString:
    """A string value wrapper."""

    _inner: PyValueString


    def __init__(self, value: str):
        self._inner = PyValueString(value)

    @classmethod
    def concrete(cls, value: str) -> "ValueString":
        """
        Create a concrete string value.

        Examples
        --------
        ```python
        >>> val = ValueString.concrete("hello")
        >>> val.is_concrete()
        True
        >>> val.get_concrete()
        'hello'
        >>>
        ```
        """
        instance = cls.__new__(cls)
        instance._inner = PyValueString.concrete(value)
        return instance

    @classmethod
    def template(cls, template: str) -> "ValueString":
        """
        Create a template string value.

        Examples
        --------
        ```python
        >>> val = ValueString.template("{{ version }}")
        >>> val.is_template()
        True
        >>> val.get_template()
        '{{ version }}'
        >>>
        ```
        """
        instance = cls.__new__(cls)
        instance._inner = PyValueString.template(template)
        return instance

    def is_concrete(self) -> bool:
        """
        Check if this is a concrete value.

        Examples
        --------
        ```python
        >>> ValueString.concrete("test").is_concrete()
        True
        >>> ValueString.template("{{ var }}").is_concrete()
        False
        >>>
        ```
        """
        return self._inner.is_concrete()

    def is_template(self) -> bool:
        """
        Check if this is a template value.

        Examples
        --------
        ```python
        >>> ValueString.template("{{ var }}").is_template()
        True
        >>> ValueString.concrete("test").is_template()
        False
        >>>
        ```
        """
        return self._inner.is_template()

    def get_concrete(self) -> Optional[str]:
        """Get the concrete value."""
        return self._inner.get_concrete()

    def get_template(self) -> Optional[str]:
        """Get the template value."""
        return self._inner.get_template()

    @classmethod
    def _from_inner(cls, inner: PyValueString) -> "ValueString":
        """Create a ValueString from a PyValueString."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance
    
    def __str__(self) -> str:
        """
        Get the string representation of the ValueString.

        Returns
        -------
        str
            The concrete value if available, otherwise the template.
        """
        return str(self._inner)


class ValueU64:
    """A U64 value wrapper."""

    _inner: PyValueU64

    @classmethod
    def concrete(cls, value: int) -> "ValueU64":
        """
        Create a concrete U64 value.

        Examples
        --------
        ```python
        >>> val = ValueU64.concrete(42)
        >>> val.is_concrete()
        True
        >>> val.get_concrete()
        42
        >>>
        ```
        """
        instance = cls.__new__(cls)
        instance._inner = PyValueU64.concrete(value)
        return instance

    @classmethod
    def template(cls, template: str) -> "ValueU64":
        """
        Create a template U64 value.

        Examples
        --------
        ```python
        >>> val = ValueU64.template("{{ build_number }}")
        >>> val.is_template()
        True
        >>> val.get_template()
        '{{ build_number }}'
        >>>
        ```
        """
        instance = cls.__new__(cls)
        instance._inner = PyValueU64.template(template)
        return instance

    def is_concrete(self) -> bool:
        """
        Check if this is a concrete value.

        Examples
        --------
        ```python
        >>> ValueU64.concrete(123).is_concrete()
        True
        >>> ValueU64.template("{{ num }}").is_concrete()
        False
        >>>
        ```
        """
        return self._inner.is_concrete()

    def is_template(self) -> bool:
        """
        Check if this is a template value.

        Examples
        --------
        ```python
        >>> ValueU64.template("{{ num }}").is_template()
        True
        >>> ValueU64.concrete(123).is_template()
        False
        >>>
        ```
        """
        return self._inner.is_template()

    def get_concrete(self) -> Optional[int]:
        """Get the concrete value."""
        return self._inner.get_concrete()

    def get_template(self) -> Optional[str]:
        """Get the template value."""
        return self._inner.get_template()

    @classmethod
    def _from_inner(cls, inner: PyValueU64) -> "ValueU64":
        """Create a ValueU64 from a PyValueU64."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance


class ConditionalRequirements:
    """A conditional requirements wrapper."""

    _inner: PyConditionalRequirements

    def __init__(self) -> None:
        self._inner = PyConditionalRequirements()

    @property
    def build(self) -> "ConditionalListPackageDependency":
        """Get the build requirements."""
        return [ItemPackageDependency._from_inner(build) for build in self._inner.build]

    @build.setter
    def build(self, value: "ConditionalListPackageDependency") -> None:
        """Set the build requirements."""
        self._inner.build = value

    @property
    def host(self) -> "ConditionalListPackageDependency":
        """Get the host requirements."""
        return [ItemPackageDependency._from_inner(host) for host in self._inner.host]

    @host.setter
    def host(self, value: "ConditionalListPackageDependency") -> None:
        """Set the host requirements."""
        self._inner.host = value

    @property
    def run(self) -> "ConditionalListPackageDependency":
        """Get the run requirements."""
        return [ItemPackageDependency._from_inner(run) for run in self._inner.run]

    @run.setter
    def run(self, value: "ConditionalListPackageDependency") -> None:
        """Set the run requirements."""
        self._inner.run = value

    @property
    def run_constraints(self) -> "ConditionalListPackageDependency":
        """Get the run constraints."""
        return [ItemPackageDependency._from_inner(run_constraint) for run_constraint in self._inner.run_constraints]

    @run_constraints.setter
    def run_constraints(self, value: "ConditionalListPackageDependency") -> None:
        """Set the run constraints."""
        self._inner.run_constraints = value

    def resolve(self, host_platform: Optional[Platform] = None) -> "PackageSpecDependencies":
        """Resolve the requirements."""
        py_platform = host_platform._inner if host_platform else None
        return PackageSpecDependencies._from_inner(self._inner.resolve(py_platform))

    @classmethod
    def _from_inner(cls, inner: PyConditionalRequirements) -> "ConditionalRequirements":
        """Create a ConditionalRequirements from a PyConditionalRequirements."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance
    
    def __str__(self) -> str:
        """
        Get the string representation of the ConditionalRequirements.

        Returns
        -------
        str
            The string representation of the build, host, run, and run constraints.
        """
        return str(self._inner)


class About:
    """An about information wrapper."""

    _inner: PyAbout

    def __init__(self) -> None:
        self._inner = PyAbout()

    @property
    def homepage(self) -> Optional[str]:
        """Get the homepage."""
        return self._inner.homepage

    @property
    def license(self) -> Optional[str]:
        """Get the license."""
        return self._inner.license

    @property
    def summary(self) -> Optional[str]:
        """Get the summary."""
        return self._inner.summary

    @property
    def description(self) -> Optional[str]:
        """Get the description."""
        return self._inner.description

    @classmethod
    def _from_inner(cls, inner: PyAbout) -> "About":
        """Create an About from a PyAbout."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance


class Extra:
    """An extra information wrapper."""

    _inner: PyExtra

    def __init__(self) -> None:
        self._inner = PyExtra()

    @classmethod
    def _from_inner(cls, inner: PyExtra) -> "Extra":
        """Create an Extra from a PyExtra."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance

    @property
    def recipe_maintainers(self) -> "ConditionalListString":
        """Get the recipe maintainers."""
        return self._inner.recipe_maintainers


class PackageSpecDependencies:
    """A package spec dependencies wrapper."""

    _inner: PyPackageSpecDependencies

    @classmethod
    def _from_inner(cls, inner: PyPackageSpecDependencies) -> "PackageSpecDependencies":
        """Create a PackageSpecDependencies from a PyPackageSpecDependencies."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance

    @property
    def host(self) -> Dict[str, str]:
        """Get the host dependencies."""
        return self._inner.host

    @property
    def run(self) -> Dict[str, str]:
        """Get the run dependencies."""
        return self._inner.run

    @property
    def run_constraints(self) -> Dict[str, str]:
        """Get the run constraints."""
        return self._inner.run_constraints

    @property
    def build(self) -> Dict[str, str]:
        """Get the build dependencies."""
        return self._inner.build


class ItemPackageDependency:
    """A package dependency item wrapper."""

    _inner: PyItemPackageDependency

    def __init__(self, name: str):
        self._inner = PyItemPackageDependency(name)


    @classmethod
    def _from_inner(cls, inner: PyItemPackageDependency) -> "ItemPackageDependency":
        """Create an ItemPackageDependency from a FFI PyItemPackageDependency."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance

    def __str__(self):
        return str(self._inner)
    
    @property
    def concrete(self) -> "PackageDependency":
        """Get the concrete package dependency."""
        return self._inner.concrete()
    
    @property
    def template(self) -> Optional[str]:
        """Get the template string if this is a template."""
        return self._inner.template()




class ItemString:
    """A package dependency item wrapper."""

    _inner: PyItemString

    @classmethod
    def _from_inner(cls, inner: PyItemString) -> "ItemString":
        """Create an ItemString from a FFI PyItemString."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance


class Source:
    """A source wrapper."""

    _inner: PySource

    @classmethod
    def url(cls, url: str) -> "Source":
        """Create a URL source."""
        instance = cls.__new__(cls)
        instance._inner = PySource.url(PyUrlSource(url, None))
        return instance

    @classmethod
    def path(cls, path: Path) -> "Source":
        """Create a path source."""
        instance = cls.__new__(cls)
        instance._inner = PySource.path(PyPathSource(str(path), None))
        return instance

    @classmethod
    def _from_inner(cls, inner: PySource) -> "Source":
        """Create a Source from a PySource."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance


class UrlSource:
    """A URL source wrapper."""

    _inner: PyUrlSource

    def __init__(self, url: str, sha: Optional[str] = None):
        self._inner = PyUrlSource(url, sha)

    @classmethod
    def _from_inner(cls, inner: PyUrlSource) -> "UrlSource":
        """Create a UrlSource from a PyUrlSource."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance

    @property
    def url(self) -> str:
        """Get the URL."""
        return self._inner.url

    @property
    def sha(self) -> Optional[str]:
        """Get the SHA."""
        return self._inner.sha


class PathSource:
    """A path source wrapper."""

    _inner: PyPathSource

    def __init__(self, path: Path, sha: Optional[str] = None):
        self._inner = PyPathSource(path, sha)

    @classmethod
    def _from_inner(cls, inner: PyPathSource) -> "PathSource":
        """Create a PathSource from a PyPathSource."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance

    @property
    def path(self) -> str:
        """Get the path."""
        return self._inner.path
