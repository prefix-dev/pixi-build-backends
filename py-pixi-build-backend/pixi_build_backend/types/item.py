from typing import List, Optional
from pixi_build_backend.pixi_build_backend import PyConditionalString, PyListOrItemString, PyVecItemPackageDependency, PyItemPackageDependency
from pixi_build_backend.types.conditional import ConditionalPackageDependency
from pixi_build_backend.types.requirements import PackageDependency

class VecItemPackageDependency(List):
    """A wrapper for a list of ItemPackageDependency."""

    _inner: PyVecItemPackageDependency

    def __init__(self, items=None):
        """Initialize with optional items."""
        if items is None:
            self._inner = PyVecItemPackageDependency()
        else:
            # Extract _inner from ItemPackageDependency objects
            inner_items = [item._inner if isinstance(item, ItemPackageDependency) else item for item in items]
            self._inner = PyVecItemPackageDependency(inner_items)

    @classmethod
    def _from_inner(cls, inner: PyVecItemPackageDependency) -> "VecItemPackageDependency":
        """Create from PyVecItemPackageDependency."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance

    def __len__(self) -> int:
        """Return the length of the list."""
        return len(self._inner)

    def __getitem__(self, index) -> "ItemPackageDependency":
        """Get item at index."""
        return ItemPackageDependency._from_inner(self._inner[index])

    def __setitem__(self, index, value):
        """Set item at index."""
        inner_value = value._inner if isinstance(value, ItemPackageDependency) else value
        self._inner[index] = inner_value

    def __delitem__(self, index):
        """Delete item at index."""
        del self._inner[index]

    def __iter__(self):
        """Return iterator."""
        for item in self._inner:
            yield ItemPackageDependency._from_inner(item)

    def __contains__(self, item: "ItemPackageDependency") -> bool:
        """Check if item is in the list."""
        inner_item = item._inner if isinstance(item, ItemPackageDependency) else item
        return inner_item in self._inner

    def append(self, item: "ItemPackageDependency"):
        """Append item to the list."""
        inner_item = item._inner
        self._inner.append(inner_item)

    def extend(self, items: List["ItemPackageDependency"]):
        """Extend the list with items."""
        inner_items = [item._inner for item in items]
        self._inner.extend(inner_items)

    def insert(self, index, item: "ItemPackageDependency"):
        """Insert item at index."""
        inner_item = item._inner
        self._inner.insert(index, inner_item)

    def remove(self, item: "ItemPackageDependency"):
        """Remove first occurrence of item."""
        inner_item = item._inner
        self._inner.remove(inner_item)

    def pop(self, index=-1):
        """Remove and return item at index."""
        return ItemPackageDependency._from_inner(self._inner.pop(index))

    def clear(self):
        """Remove all items."""
        self._inner.clear()

    def index(self, item: "ItemPackageDependency", start=0, stop=None):
        """Return index of first occurrence of item."""
        inner_item = item._inner
        if stop is None:
            return self._inner.index(inner_item, start)
        else:
            return self._inner.index(inner_item, start, stop)

    def count(self, item) -> int:
        """Return count of occurrences of item."""
        inner_item = item._inner if isinstance(item, "ItemPackageDependency") else item
        return self._inner.count(inner_item)

    def reverse(self):
        """Reverse the list in place."""
        self._inner.reverse()

    def sort(self, key=None, reverse=False):
        """Sort the list in place."""
        if key is None:
            self._inner.sort(reverse=reverse)
        else:
            self._inner.sort(key=key, reverse=reverse)

    def copy(self):
        """Return a shallow copy."""
        return VecItemPackageDependency._from_inner(self._inner.copy())

    def __eq__(self, other) -> bool:
        """Check equality."""
        if isinstance(other, VecItemPackageDependency):
            return self._inner == other._inner
        return False

    def __str__(self) -> str:
        """Return string representation."""
        return str(self._inner)
    

class ItemPackageDependency:
    """A package dependency item wrapper."""

    _inner: PyItemPackageDependency

    def __init__(self, name: str):
        self._inner = PyItemPackageDependency(name)

    @classmethod
    def new_from_conditional(
        cls,
        conditional: ConditionalPackageDependency
    ) -> "ItemPackageDependency":
        new_class = cls.__new__(cls)
        new_class._inner = PyItemPackageDependency.new_from_conditional(conditional._inner)
        return new_class


    @classmethod
    def _from_inner(cls, inner: PyItemPackageDependency) -> "ItemPackageDependency":
        """Create an ItemPackageDependency from a FFI PyItemPackageDependency."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance

    def __str__(self):
        return str(self._inner)
    
    @property
    def concrete(self) -> "PackageDependency":
        """Get the concrete package dependency."""
        return PackageDependency._from_inner(self._inner.concrete())
    
    @property
    def template(self) -> Optional[str]:
        """Get the template string if this is a template."""
        return self._inner.template()
    
    @property
    def conditional(self) -> Optional[ConditionalPackageDependency]:
        """Get the conditional string if this is a conditional."""
        return self._inner.conditional()