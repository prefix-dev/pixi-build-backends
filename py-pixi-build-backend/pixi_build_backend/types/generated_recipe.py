from typing import Any, Dict, Optional, Protocol, List
from pathlib import Path
from pixi_build_backend.types.intermediate_recipe import IntermediateRecipe
from pixi_build_backend.pixi_build_backend import PyGeneratedRecipe, PyGenerateRecipe
from pixi_build_backend.types.platform import Platform
from pixi_build_backend.types.project_model import ProjectModelV1
from pixi_build_backend.types.python_params import PythonParams


class GeneratedRecipe:
    """A generated recipe wrapper."""

    _inner: PyGeneratedRecipe

    def __init__(self) -> None:
        self._inner = PyGeneratedRecipe()

    @classmethod
    def from_model(cls, model: ProjectModelV1, manifest_root: Path) -> "GeneratedRecipe":
        """Create a GeneratedRecipe from a ProjectModelV1."""
        instance = cls()
        instance._inner = PyGeneratedRecipe().from_model(model._inner, manifest_root)
        return instance

    @property
    def recipe(self) -> IntermediateRecipe:
        """Get the recipe."""
        return IntermediateRecipe._from_inner(self._inner.recipe)

    @property
    def metadata_input_globs(self) -> List[str]:
        """Get the metadata input globs."""
        return self._inner.metadata_input_globs

    @property
    def build_input_globs(self) -> List[str]:
        """Get the build input globs."""
        return self._inner.build_input_globs

    def __repr__(self) -> str:
        return self._inner.__repr__()

    def _into_py(self) -> PyGeneratedRecipe:
        """
        Converts this object into a type that can be used by the Rust code.
        """
        return self._inner


class GenerateRecipeProtocol(Protocol):
    """
    Protocol for generating recipes.
    This should be implemented by the Python generator.
    """

    def generate_recipe(
        self,
        model: ProjectModelV1,
        config: Dict[str, Any],
        manifest_path: str,
        host_platform: Platform,
        python_params: Optional[PythonParams] = None,
    ) -> GeneratedRecipe:
        """Generate a recipe for a Python package."""
        ...


class GenerateRecipe:
    """Protocol for generating recipes."""

    _inner: PyGenerateRecipe

    def __init__(self, instance: GenerateRecipeProtocol):
        self._inner = PyGenerateRecipe(instance)

    def _into_py(self) -> PyGenerateRecipe:
        """
        Converts this object into a type that can be used by the Rust code.
        """
        return self._inner
