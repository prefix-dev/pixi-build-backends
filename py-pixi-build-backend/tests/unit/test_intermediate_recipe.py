from pathlib import Path
from typing import Any
from pixi_build_backend.types.intermediate_recipe import IntermediateRecipe


def test_from_yaml(snapshot: Any) -> None:
    yaml_file = Path(__file__).parent.parent / "data" / "boltons_recipe.yaml"
    yaml_content = yaml_file.read_text()

    recipe = IntermediateRecipe.from_yaml(yaml_content)

    assert snapshot == recipe.to_yaml()
