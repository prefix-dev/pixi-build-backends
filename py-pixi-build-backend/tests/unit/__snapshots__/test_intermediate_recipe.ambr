# serializer version: 1
# name: test_from_yaml
  '''
  context:
    version: 23.0.0
  package:
    name: boltons-with-extra
    version: ${{ version }}
  source:
  - url: https://github.com/mahmoud/boltons/archive/refs/tags/${{ version }}.tar.gz
    sha256: 9b2998cd9525ed472079c7dd90fbd216a887202e8729d5969d4f33878f0ff668
  build:
    number: null
    script:
      content:
      - python -m pip install . --no-deps -vv
      env: {}
      secrets: []
    noarch: python
  requirements:
    build: []
    host:
    - python
    - pip
    - setuptools
    run:
    - pip
    run_constraints: []
  tests: []
  about:
    homepage: null
    license: BSD-3-Clause
    license_file: LICENSE
    summary: null
    description: null
    documentation: null
    repository: null
  extra: null
  
  '''
# ---
